building_sim_plugins/building_plugins_common/src/crowd_simulator_common.cpp

```diff
@@ -0,0 +1,310 @@
+#include <fstream>
+#include <cmath>
+
+#include <building_sim_common/crowd_simulator_common.hpp>
+
+namespace crowd_simulator {
+
+//================================================================
+void MengeHandle::SetSimTimeStep(float simTimeStep)
+{
+  this->_simTimeStep = simTimeStep;
+}
+
+float MengeHandle::GetSimTimeStep()
+{
+  return this->_simTimeStep;
+}
+
+size_t MengeHandle::GetAgentCount()
+{
+  if (this->_agentCount == 0)
+  {
+    this->_agentCount = this->_sim->getNumAgents();
+  }
+  return this->_agentCount;
+}
+
+void MengeHandle::SimStep()
+{
+  this->_sim->step();
+}
+
+AgentPtr MengeHandle::GetAgent(size_t id)
+{
+  return AgentPtr(this->_sim->getAgent(id));
+}
+
+std::string MengeHandle::_ResourceFilePath(const std::string& relativePath)
+const
+{
+
+  std::string fullPath = this->_resourcePath + "/" + relativePath;
+  std::cout << "Finding resource file: " << fullPath << std::endl;
+  std::ifstream ifile(fullPath);
+  if (!static_cast<bool>(ifile))
+  {
+    std::cerr << "File not found! " << fullPath << std::endl;
+    assert(static_cast<bool>(ifile));
+  }
+  std::cout << "Found." << std::endl;
+  return fullPath;
+}
+
+bool MengeHandle::_LoadSimulation()
+{
+  Menge::SimulatorDB simDB;
+  Menge::PluginEngine::CorePluginEngine engine(&simDB);
+
+  std::cout << "Start CrowdSimulator initializing [Menge]..." << std::endl;
+
+  this->_sim = std::shared_ptr<Menge::Agents::SimulatorInterface> (
+    simDB.getDBEntry("orca")->getSimulator(
+      this->_agentCount,
+      this->_simTimeStep,
+      0,
+      std::numeric_limits<float>::infinity(),
+      this->_behaviorFile,
+      this->_sceneFile,
+      "",
+      "",
+      false)
+  );
+
+  if (this->_sim)
+  {
+    std::cout << std::endl << "Crowd Simulator initialized success [Menge]. " <<
+      std::endl;
+    return true;
+  }
+
+  std::cout << "Error in provided navmesh. Menge simulator initialized false." << std::endl;
+
+  return false;
+}
+
+//============================================
+
+ModelTypeDatabase::Record* ModelTypeDatabase::Get(const std::string& typeName)
+{
+  if (this->_records.find(typeName) == this->_records.end())
```

Comment:
A minor optimization that can be done here,
```
auto it = this->_records.find(typeName);
if (it == _records.end())
{
  return nullptr;
}
return it->second;
```

Reply:
Now that I look at it it should also be possible to make the function const by adding @aaronchongth optimization and removing the square bracket operator?

Reply:
Good idea. So the [] operator is usually not preferred? I should use iterator as much as possible right?

Reply:
I will say it is still highly dependent on the use-case, @luca-della-vedova yes the method can be a `const` qualified, good catch